.gitlab-ci.yml

```diff
@@ -2,6 +2,30 @@ stages:
   - build
   - release
 
+.setup-environment:
+  image: 'cimg/go:1.16'
+  variables:
+    GOPATH: "$CI_PROJECT_DIR/.go"
+  before_script:
+    - mkdir -p .go
+  script:
+    - make install-tools
+  cache:
+    paths:
+      - .go/pkg/mod
+      - .go/bin
+
+compile:
+  extends: .setup-environment
+  stage: build
+  parallel:
+    matrix:
+      - TARGET: [binaries-darwin_amd64, binaries-linux_amd64, binaries-linux_arm64, binaries-windows_amd64]
+  script: make $TARGET
```

Comment:
won't this script override `- make install-tools`? shouldn't `make install-tools` go to `before_script` or another stage prior to build?

Reply:
I'm new to gitlab, so still learning.  I guess I can try moving `make install-tools` to a separate job in the `.pre` stage?

Reply:
Yeah, probably `.pre` stage is better. `extends:` is just a way to reuse the same config parts between tasks